### DISCUSSION TITLE
RDFPandas, providing RDF support for Pandas

### DISCUSSION THREAD

**Discussion by ghost** [RDFLib/rdflib#1734] [Show and tell]
### [RdfPandas](https://github.com/cadmiumkitty/rdfpandas#rdfpandas)

RdfPandas is a module providing RDF support for Pandas. It consists of two simple functions for `Graph` to `DataFrame` conversion and `DataFrame` to `Graph` conversion.

#### [Creating RDF from DataFrame](https://github.com/cadmiumkitty/rdfpandas#creating-dataframe-from-rdf)
As of version 1.1.0 NamespaceManager can be supplied to `rdflib.to_graph` for conversion to Graph.

```py
from rdfpandas.graph import to_graph
import pandas as pd
import rdflib

df = pd.read_csv('to_graph_test.csv', index_col = '@id', keep_default_na = False)
namespace_manager = NamespaceManager(Graph())
namespace_manager.bind('skos', SKOS)
namespace_manager.bind('rdfpandas', Namespace('http://github.com/cadmiumkitty/rdfpandas/'))
g = to_graph(df, namespace_manager)
s = g.serialize(format = 'turtle')
```
#### Creating DataFrame from RDF
```py
from rdfpandas.graph import to_dataframe
import pandas as pd
import rdflib

g = rdflib.Graph()
g.parse('to_df_test.ttl', format = 'ttl')
df = to_dataframe(g)
df.to_csv('test.csv', index = True, index_label = "@id")
```

#### Gotchas

No special effort is made for dealing with types, so please be aware of Pandas features such as https://pandas.pydata.org/pandas-docs/stable/user_guide/gotchas.html#support-for-integer-na that may result in surprising RDF statements like `"10.0"^^<xsd:integer>`.

**Comment by davidshumway** [comment]
Very interesting. I'm curious what the equivalent functionality is for sending the triples in a graph to a DataFrame? As a first try, I'd think something like this?

```python
g = Graph()
g.parse('test.ttl')
p = pd.DataFrame()
for stmt in g:
  p = pd.concat([p, pd.DataFrame({'s': [stmt[0]], 'p': [stmt[1]], 'o': [stmt[2]]})])
```

For a graph with 10,000 triples this is pretty efficient but for larger graphs it seems to take longer.
